### DISCUSSION TITLE
Async version of SessionState.InvokeCommand.InvokeScript?

### DISCUSSION THREAD

**Discussion by potatoqualitee** [PowerShell/PowerShell#18429] [Q&A] [answered]
Hello,
I am again trying to get some blazing fast speeds from our imports. So far, the fastest way to do it is to use C# to Import one large file with all commands via SessionState.InvokeCommand.InvokeScript.

But when I was looking for even faster techniques to import (StreamWriter, ReadAllText), etc, someone suggested `Parallel.ForEach` but I have not gotten that to work. I dont even know if it will be faster in the end but I'd like to try.

When I use SessionState.InvokeCommand.InvokeScript. with the Parallel, it bails on whatever my max parallellism is set to saying an error has occurred then when I dig deeper, it's the tokenizer losing its mind. Basically, it looks like a few threads collide and make one big jumbled code block.

This is basically how I do it:

Get Directory of 650+ functions, Parallel.ForEach through the files, then SessionState.InvokeCommand.InvokeScript the text from that file.

If I make each thread sleep for 1 second after, it works. Which makes me believe it's an async issue. I'd love to know if there's an async way to do this. @SeeminglyScience would you know? I saw you participate on a few threads that suggested you would.

**Comment by SeeminglyScience** [comment, marked as answer]
Pretty much all of PowerShell's state is thread local and cannot be accessed (and especially not mutated) from another thread. The sleep is unfortunately just making race conditions a little less likely.

I know this isn't strictly about dbatools, but has the project tried [ModuleBuilder](https://github.com/PoshCode/ModuleBuilder)? Specifically, the function [`Build-Module`](https://github.com/PoshCode/ModuleBuilder#build-module) that would be done as a build step before publishing. That's usually what I recommend to folks running into script based import performance issues

**Reply by SeeminglyScience** [reply]
That said... `Parser.ParseFile` is thread safe. You could maybe do something like

```csharp
// ConcurrentDictionary probably doesn't make sense. ConcurrentBag would be better
ConcurrentDictionary<string, ScriptBlock> scriptblocks = new();
Parallel.ForEach(() =>
{
    scriptblocks.GetOrAdd(currentFile, file =>
    {
        // probably want to handle errors here
        var ast = (ScriptBlockAst)Parser.ParseFile(file, out _, out _);
        return ast?.GetScriptBlock();
    }
});

// The actual invoking *must* be done on the pipeline thread
foreach (var sb in scriptblocks.Values)
{
    SessionState.InvokeCommand.InvokeScript(useLocalScope: false, sb, input: null);
}
```

Though for sure check out `Build-Module` still

Also if any of the individual files have a `using module` statement the perf is gonna tank hard as that's cached in thread local storage iirc

**Reply by potatoqualitee** [reply]
That is a nice module! We currently use `Help-Out` to do something similar. It both exports the help to an xml file and combines all scripts into one file. I believe that separating the help shouldn't make it faster but my tests repeatedly showed it did so we started building it that way.

I'm pumped about testing this code! Thank you so much, gonna go try it.

**Reply by potatoqualitee** [reply]
So here's what it looks like in the older C# style and a bag as written by a PowerSheller:

```csharp
            string[] files = Directory.GetFiles(Path);
            ConcurrentBag<ScriptBlock> scriptblocks = new ConcurrentBag<ScriptBlock>();
            Parallel.ForEach(files, file =>
            {
                Token[] tokens;
                ParseError[] errors;
                var ast = (ScriptBlockAst)Parser.ParseFile(file, out tokens, out errors);
                scriptblocks.Add(ast.GetScriptBlock());
            });

            // The actual invoking *must* be done on the pipeline thread
            foreach (ScriptBlock sb in scriptblocks)
            {
                SessionState.InvokeCommand.InvokeScript(false, sb, null, null);
            }
```

As suspected, it's indeed slower, still at 6+ seconds. So there's no way to get that InvokeScript in the Parallel.ForEach eh? Next up, I'm going to divide the one big file into 6, set the multithreading to 7 and see what comes of it 😅  Thanks again for the response.

**Reply by SeeminglyScience** [reply]
> So here's what it looks like in the older C# style and a bag as written by a PowerSheller:

Very nice ❤️ 

> So there's no way to get that InvokeScript in the Parallel.ForEach eh?

Yeah, even if you find a way to make it work it'll be really dangerous. Only saying that because my reaction when someone says that is always "impossible?? we'll see about that!" 😁 but seriously, a lot of state corruption can and will eventually happen. So instead of saying it's impossible I'd rather say it's directly at odds with PowerShell's state management infrastructure.

---

Just a side note that if you intend to ship the C# as an `Add-Type` statement, that's probably gonna kill any perf bonus you get from it in enterprise environments. Places with real time monitoring make the `Add-Type` take a few literal seconds, it's pretty rough

**Reply by potatoqualitee** [reply]
I have indeed noticed what you're saying about the state corruption and was bummed!

Yesterday, I tried the safer way with 2, 4 and 6 files and perf was worse. But, I took a note from your use of ScriptBlock and used the ScriptBlock way instead of the text way and I swear, I got an average of 200ms of improvements and up to 500ms on some runs! Now I'm going to look at performing improvements in other parts of the module loading.

I believe we use Import-Module for most of our DLL imports. We have to import a lot of assemblies as we rely heavily on SQL Management Objects, SqlClient and DacFX and there's no getting around that, so we just accept that dbatools will load slower in secure-r environments. As long as I can get it fast on a regular workstation or my Linux server, I'm okay with that. Loading all the SMO adds about 200ms I believe. So on bare-metal Linux, the shortest import times we can get is probably 1 second on average. Windows seems to be 2.6 seconds so far. Microsoft's module has 1/6th the commands, are C# cmdlets and loads at a whopping 250ms, I'm so jealous 😭 

Do you know of any way to make PowerShell load the SMO libraries when it needs them? Right now, I have to explicitly import about 10 DLLs in addition to the dbatools library. Well, I just looked and it's 4 DLLs in Linux and 11 DLLs in Windows. In Windows, I also have to do a bunch of redirections that I don't have to do in Core.

```powershell
if ($PSVersionTable.PSEdition -ne "Core") {
    $dir = (Join-Path $script:libraryroot "lib\").Replace('\', '\\')

    if (-not ("Redirector" -as [type])) {
        $source = @"
            using System;
            using System.Linq;
            using System.Reflection;
            using System.Text.RegularExpressions;

            public class Redirector
            {
                public Redirector()
                {
                    this.EventHandler = new ResolveEventHandler(AssemblyResolve);
                }

                public readonly ResolveEventHandler EventHandler;

                protected Assembly AssemblyResolve(object sender, ResolveEventArgs e)
                {
                    string[] dlls = {
                        "System.Runtime.CompilerServices.Unsafe",
                        "System.Resources.Extensions",
                        "Microsoft.SqlServer.ConnectionInfo",
                        "Microsoft.SqlServer.Smo",
                        "Microsoft.Identity.Client",
                        "System.Diagnostics.DiagnosticSource",
                        "Microsoft.IdentityModel.Abstractions",
                        "Microsoft.Data.SqlClient",
                        "System.Configuration.ConfigurationManager",
                        "Microsoft.SqlServer.Replication",
                        "Microsoft.SqlServer.Rmo",
                        "System.Private.CoreLib"
                    };

                    var name = new AssemblyName(e.Name);
                    var assemblyName = name.Name.ToString();
                    foreach (string dll in dlls)
                    {
                        if (assemblyName == dll)
                        {
                            string filelocation = "$dir" + dll + ".dll";
                            //Console.WriteLine(filelocation);
                            return Assembly.LoadFrom(filelocation);
                        }
                    }

                    foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies())
                    {
                        var info = assembly.GetName();
                        if (info.FullName == e.Name) {
                            return assembly;
                        }
                    }
                    return null;
                }
            }
"@

        $null = Add-Type -TypeDefinition $source
    }

    try {
        $redirector = New-Object Redirector
        [System.AppDomain]::CurrentDomain.add_AssemblyResolve($redirector.EventHandler)
    } catch {
        # unsure
    }
}

$dll = Join-Path $script:libraryroot "lib\dbatools.dll"

try {
    Import-Module "$dll"
} catch {
    throw "Couldn't import dbatools.dll | $PSItem"
}

if ($PSVersionTable.PSEdition -ne "Core") {
    [System.AppDomain]::CurrentDomain.remove_AssemblyResolve($onAssemblyResolveEventHandler)
}
```

Glad I got it working reliably tho -- I've needed to for years.

**Reply by SeeminglyScience** [reply]
> Do you know of any way to make PowerShell load the SMO libraries when it needs them? Right now, I have to explicitly import about 10 DLLs in addition to the dbatools library.

Soooo it depends on how the types within are used. As soon as they're used in external metadata (output type, parameter type, etc) or in a class, they pretty much need to be loaded up front. Otherwise you can sort of get away with a check in the `begin` block.

That said, you might be able to load (but *not* import) the assemblies in a different thread, then later on in the process import them. It's mostly only PowerShell's state that is thread local.

So something like this psuedo code:

```powershell
$job = Do-AsyncSomehow {
    [Assembly]::Load('MyAssembly1')
    [Assembly]::Load('MyAssembly2')
    [Assembly]::Load('MyAssembly3')
}

# Do some other tasks here

$job | wait
Import-Module -Assembly ([TypeFromAssembly1].Assembly)
Import-Module -Assembly ([TypeFromAssembly2].Assembly)
Import-Module -Assembly ([TypeFromAssembly3].Assembly)
```

**Reply by potatoqualitee** [reply]
OH HECK YES! This looks awesome like a wonderful solution, can't wait to try it. I'll let you know how it goes.
